### PR TITLE
Remove gun-specific code from stub to improve client adapter flexibility

### PR DESCRIPTION
**Pull request by beligante**
I'm working on this issue: https://github.com/elixir-grpc/grpc/issues/242

And while working on that, I've found that `stub.ex` has a lot of `gun` specific code and that should live inside the adapter. 

This doesn't contains drastic code changes, I'm simply moving a small piece of code around to keep on working on adapter code.

My Idea here is to create an layer of abstraction, where each adapter implements it's own version of receiving data. For `gun` we have the mechanics of receive the headers first and then the payload. 

When working with Mint that's not the case. See [`stream/2` docs](https://hexdocs.pm/mint/Mint.HTTP.html#stream/2) 
TL;DR: The message we receive in the process might contain headers and parts of the payload (in the same message). So having a function that receives the headers and the body doesn't make to much sense.

### CODE REVIEW

File: lib/grpc/client/adapter.ex [hunk 1]
```diff
@@ -17,13 +17,10 @@ defmodule GRPC.Client.Adapter do
...
-
-  @callback recv_data_or_trailers(
-              stream :: map(),
-              trailers_or_metadata :: map(),
-              opts :: keyword()
-            ) ::
-              {:data, binary()} | {:trailers, binary()} | {:error, GRPC.RPCError.t()}
+  @callback receive_data(stream :: Stream.t(), opts :: keyword()) ::
+              {:ok, struct()}
+              | {:ok, struct(), map()}
```
Comment (polvalente) [inline lib/grpc/client/adapter.ex:22]: what struct can be returned here?

File: lib/grpc/client/adapter.ex [hunk 2]
```diff
@@ -17,13 +17,10 @@ defmodule GRPC.Client.Adapter do
...
+              | {:ok, Enumerable.t()}
+              | {:ok, Enumerable.t(), map()}
```
Comment (polvalente) [inline lib/grpc/client/adapter.ex:24]: what is the enumerable here? `name :: type` notation could be useful

File: lib/grpc/client/adapters/gun.ex [hunk 3]
```diff
@@ -135,7 +135,42 @@ defmodule GRPC.Client.Adapters.Gun do
...
+        %{server_stream: true, channel: %{adapter_payload: adapter_payload}, payload: payload} =
+          stream,
+        opts
+      ) do
+    with {:ok, headers, is_fin} <- recv_headers(adapter_payload, payload, opts) do
+      case is_fin do
+        :fin -> []
+        :nofin -> response_stream(stream, opts)
+      end
+      |> then(&{:ok, &1})
```
Comment (polvalente) [inline lib/grpc/client/adapters/gun.ex:148]:
```suggestion
        :fin -> {:ok, []}
        :nofin -> {:ok, response_stream(stream, opts)}
      end
```

for readability

File: lib/grpc/client/adapters/gun.ex [hunk 4]
```diff
@@ -135,7 +135,42 @@ defmodule GRPC.Client.Adapters.Gun do
...
+      |> maybe_return_headers(opts[:return_headers], %{headers: headers})
```
Comment (polvalente) [inline lib/grpc/client/adapters/gun.ex:149]: I don't quite like `maybe_*` naming. I'd rather use plain `return_headers` or avoid the function altogether in favor of an in-place if

Comment (polvalente) [reply lib/grpc/client/adapters/gun.ex:149]: It also doesn't just return headers because you use it to return a map which can also contain trailers

File: lib/grpc/client/adapters/gun.ex [hunk 5]
```diff
@@ -287,4 +321,146 @@ defmodule GRPC.Client.Adapters.Gun do
...
+        end
+
+      error = {:error, _} ->
+        {error, %{buffer: <<>>, fin: true, fin_resp: nil}}
+    end
+  end
+
+  defp read_stream(%{buffer: buffer, need_more: false, response_mod: res_mod, codec: codec} = s) do
+    case GRPC.Message.get_message(buffer) do
+      # TODO
```
Comment (polvalente) [inline lib/grpc/client/adapters/gun.ex:409]: what's "TODO" here?

File: lib/grpc/client/adapters/gun.ex [hunk 6]
```diff
@@ -287,4 +321,146 @@ defmodule GRPC.Client.Adapters.Gun do
...
+          end
+
+        case GRPC.Message.from_data(%{compressor: compressor}, body) do
+          {:ok, msg} ->
+            {:ok, codec.decode(msg, res_mod)}
+
+          err ->
+            err
+        end
+
```
Comment (polvalente) [inline lib/grpc/client/adapters/gun.ex:451]: the nesting here seems too much. Perhaps this branch can be a separate defp?

**Review by polvalente — COMMENTED** [review]
This looks good! Some minor nitpicks and we're good to go

File: lib/grpc/client/adapter.ex (hunk 2 above)
Comment (beligante) [reply lib/grpc/client/adapter.ex:24]: In this case the enumerable refers to an `Elixir.Stream` mostly used for streamed RPCs

File: lib/grpc/client/adapters/gun.ex (hunk 5 above)
Comment (beligante) [reply lib/grpc/client/adapters/gun.ex:409]: IDK to be honest lol
This part was a copy and paste from `stub.ex` code. It's comes from the original author. Seems unnecessary. I'll remove this one

File: lib/grpc/client/adapters/gun.ex (hunk 6 above)
Comment (beligante) [reply lib/grpc/client/adapters/gun.ex:451]: yeah! Agreed! I'll refactor this

File: lib/grpc/client/adapter.ex (hunk 1 above)
Comment (beligante) [reply lib/grpc/client/adapter.ex]: Added a minor doc and a new type in `GRPC.Stub` to improve this.

File: lib/grpc/client/adapters/gun.ex [hunk 7]
```diff
@@ -135,7 +135,37 @@ defmodule GRPC.Client.Adapters.Gun do
   end
 
   @impl true
-  def recv_headers(%{conn_pid: conn_pid}, %{stream_ref: stream_ref}, opts) do
+  def receive_data(
+        %{server_stream: true, channel: %{adapter_payload: adapter_payload}, payload: payload} =
+          stream,
+        opts
+      ) do
```
Comment (polvalente) [inline lib/grpc/client/adapters/gun.ex:142]: Let's split this match, leaving only the flow-control at the function head
```suggestion
        %{server_stream: true} = stream,
        opts
      ) do
```

And then match to extract the data below. Same for the other clause

**Review by polvalente — APPROVED** [review]
